Add DebugBuilder.md (#156)
Babylonpartners/ios-playbook: Cookbook/Technical-Documents/DebugBuilder.md

Debug Builder
======================

[\[CNSMR\-1602\] Add \`DebugAuthenticatedRootBuilder\` for faster UI debugging iteration by inamiy · Pull Request \#7645 · Babylonpartners/babylon\-ios](https://github.com/Babylonpartners/babylon-ios/pull/7645)

Above PR introduces `DebugAuthenticatedRootBuilder` that allows to directly show a target view builder after authentication.

This feature improves overall UI debugging iteration speed.

## How to make a debug-builder

1. Add a new case in `enum DebugBuilderPattern`
2. Create a (mock) builder in `DebugAuthenticatedRootBuilder`

> 3. Update `Settings.bundle/Root.plist` by adding a new case key

Reviewer: Need to also update `Babylon/Brand/babylon/Settings.bundle/Root.plist`

Author: Updated in 2a1b5da .

> 4. Set `Settings.bundle`'s userDefaults flag via `Settings.app`

Reviewer: ```suggestion
4. Build and launch your app at least once after those changes, so that the Settings app can now discover the new values in the `Root.plist`
4. Set `Settings.bundle`'s userDefaults flag via `Settings.app`
```

(Note that you could also use `-debugBuilder YourCaseName` in your launch parameters in your Scheme to override those User Defaults at launch time. But might not be a good idea to suggest that since people risk forgetting to remove those and they would end up being commited)

5. Launch the app

## Example

`LocalFeatureSwitches.swift`

```diff
 public enum DebugBuilderPattern: String {

>      case none

Reviewer: Update to the new name chosen once https://github.com/Babylonpartners/babylon-ios/pull/7645/files#r293449642 has been adressed

Author: ```suggestion
     case disabled
```

     ...

     func make(
         session: MainUserSession,
         authenticatedPushEvents: Signal<PushEvent, NoError>,
         authenticatedApplicationInvocationEvents: Signal<ApplicationInvocation, NoError>,
         lifecycleEvents: Signal<LifecycleEvent, NoError>
     ) -> UIViewController? {

         switch LocalFeatureSwitches.debugBuilder {
         ...

+        // NOTE: Implement a new mock builder here.
+        case .eligibilityCheckBuilder:
+            let builder = EligibilityCheckBuilder(dependencies: dependencies, session: session)
+            return DesignLibrary.viewControllersBuilder().navigationController(dismissalStyle: .pop) { (navigation, modal) -> UIViewController in
+                return builder.make(
+                    navigation: navigation,
+                    modal: modal,
+                    presenting: modal,
+                    root: modal,
+                    switchFlow: NHSSwitchFlow(
+                        flow: modal,
+                        rootViewController: UIApplication.shared.delegate?.window??.rootViewController!
+                    )
+                )
+            }
+    }
 }
```

> `Settings.bundle/Root.plist`

Reviewer: ```suggestion
`Settings.bundle/Root.plist` AND `Babylon/Brand/babylon/Settings.bundle/Root.plist`
```

Author: Writing a specific path is easy to get changed, so I decided to add a note comment instead in 2a1b5da .

Debug Builder
======================

[\[CNSMR\-1602\] Add \`DebugAuthenticatedRootBuilder\` for faster UI debugging iteration by inamiy · Pull Request \#7645 · Babylonpartners/babylon\-ios](https://github.com/Babylonpartners/babylon-ios/pull/7645)

> Above PR introduces `DebugAuthenticatedRootBuilder` that allows to directly show a target view builder after authentication.

Reviewer 2: ```suggestion
The above PR introduces `DebugAuthenticatedRootBuilder` which allows to directly show a target view builder after authentication.
```

> This feature improves overall UI debugging iteration speed.

Reviewer 2: ```suggestion
This feature improves the overall iteration speed for UI debugging.
```
